synapse/push/mailer.py

```diff
@@ -0,0 +1,385 @@
+            # want the generated-from-names one here otherwise we'll
+            # end up with, "new message from Bob in the Bob room"
+            room_name = calculate_room_name(
+                state_by_room[room_id], user_id, fallback_to_members=False
+            )
+
+            my_member_event = state_by_room[room_id][("m.room.member", user_id)]
+            if my_member_event.content["membership"] == "invite":
+                inviter_member_event = state_by_room[room_id][
+                    ("m.room.member", my_member_event.sender)
+                ]
+                inviter_name = name_from_member_event(inviter_member_event)
+
+                if room_name is None:
+                    return INVITE_FROM_PERSON % (inviter_name,)
+                else:
+                    return INVITE_FROM_PERSON_TO_ROOM % (inviter_name, room_name)
+
+            sender_name = None
+            if len(notifs_by_room[room_id]) == 1:
+                # There is just the one notification, so give some detail
+                event = notif_events[notifs_by_room[room_id][0]["event_id"]]
+                if ("m.room.member", event.sender) in state_by_room[room_id]:
+                    state_event = state_by_room[room_id][("m.room.member", event.sender)]
+                    sender_name = name_from_member_event(state_event)
+
+                if sender_name is not None and room_name is not None:
+                    return MESSAGE_FROM_PERSON_IN_ROOM % (sender_name, room_name)
+                elif sender_name is not None:
+                    return MESSAGE_FROM_PERSON % (sender_name,)
+            else:
+                # There's more than one notification for this room, so just
+                # say there are several
+                if room_name is not None:
+                    return MESSAGES_IN_ROOM % (room_name,)
+                else:
+                    # If the room doesn't have a name, say who the messages
+                    # are from explicitly to avoid, "messages in the Bob room"
+                    sender_ids = list(set([
+                        notif_events[n['event_id']].sender
+                        for n in notifs_by_room[room_id]
+                    ]))
+
+                    return MESSAGES_FROM_PERSON % (
+                        descriptor_from_member_events([
+                            state_by_room[room_id][("m.room.member", s)]
+                            for s in sender_ids
+                        ])
+                    )
+        else:
+            # Stuff's happened in multiple different rooms
+            return MESSAGES_IN_ROOMS
+
+    def make_room_link(self, room_id):
+        return "https://matrix.to/%s" % (room_id,)
+
+    def make_notif_link(self, notif):
+        return "https://matrix.to/%s/%s" % (
+            notif['room_id'], notif['event_id']
+        )
+
+    def make_unsubscribe_link(self):
+        return "https://vector.im/#/settings"  # XXX: matrix.to
```

Comment:
Would very much prefer a direct link to an actual page with a "Unsubscribe by clicking here" button, without needing to log in.